src/backend/renderer/element/utils.rs

```diff
@@ -0,0 +1,502 @@
+        self.element.src()
+    }
+
+    fn geometry(
+        &self,
+        scale: crate::utils::Scale<f64>,
+    ) -> crate::utils::Rectangle<i32, crate::utils::Physical> {
+        let mut element_geometry = self.element.geometry(scale);
+        // First we make the element relative to the origin
+        element_geometry.loc -= self.origin;
+        // Then we scale it by our scale
+        element_geometry = element_geometry.to_f64().upscale(self.scale).to_i32_round();
+        // At last we move it back to the origin
+        element_geometry.loc += self.origin;
+        element_geometry
+    }
+
+    fn transform(&self) -> crate::utils::Transform {
+        self.element.transform()
+    }
+
+    fn damage_since(
+        &self,
+        scale: crate::utils::Scale<f64>,
+        commit: Option<crate::backend::renderer::utils::CommitCounter>,
+    ) -> Vec<crate::utils::Rectangle<i32, crate::utils::Physical>> {
+        self.element
+            .damage_since(scale, commit)
+            .into_iter()
+            .map(|rect| rect.to_f64().upscale(self.scale).to_i32_up())
+            .collect::<Vec<_>>()
+    }
+
+    fn opaque_regions(
+        &self,
+        scale: crate::utils::Scale<f64>,
+    ) -> Vec<crate::utils::Rectangle<i32, crate::utils::Physical>> {
+        self.element
+            .opaque_regions(scale)
+            .into_iter()
+            .map(|rect| rect.to_f64().upscale(self.scale).to_i32_round())
+            .collect::<Vec<_>>()
+    }
+}
+
+impl<R: Renderer, E: RenderElement<R>> RenderElement<R> for RescaleRenderElement<E> {
+    fn draw<'a>(
+        &self,
+        frame: &mut <R as Renderer>::Frame<'a>,
+        src: crate::utils::Rectangle<f64, crate::utils::Buffer>,
+        dst: crate::utils::Rectangle<i32, crate::utils::Physical>,
+        damage: &[crate::utils::Rectangle<i32, crate::utils::Physical>],
+        log: &slog::Logger,
+    ) -> Result<(), <R as Renderer>::Error> {
+        self.element.draw(frame, src, dst, damage, log)
+    }
+
+    fn underlying_storage(&self, renderer: &R) -> Option<super::UnderlyingStorage<'_, R>> {
+        self.element.underlying_storage(renderer)
+    }
+}
+
+/// A element that allows to crop another element
+#[derive(Debug)]
+pub struct CropRenderElement<E> {
+    element: E,
+    src: Rectangle<f64, Buffer>,
+    crop: Rectangle<i32, Physical>,
+    geometry: Rectangle<i32, Physical>,
+}
+
+impl<E: Element> CropRenderElement<E> {
+    /// Create a cropping render element for an existing element
+    ///
+    /// The crop rect is expected to be relative to the same origin the element is relative to.
+    /// It can extend outside of the element geometry but the resulting geometry will always
+    /// be equal or smaller than the geometry of the original element.
+    ///
+    /// The scale is used to calculate the scaling from physical space to buffer space and must
+    /// match the scale provided to the [`Element`] functions.
```

Comment:
*must* sounds very strong. What happens if we pass a scale of 2 (like the underlying element), but then render a cropped window on outputs of scales 1 and 2? Then this description is ambiguous. 

We should rather document what happens if the scale does not match the buffer scale. Or if possible get rid of the scale and specify the `crop_rect` in `Buffer`-coordinate space?

Reply:
> _must_ sounds very strong. What happens if we pass a scale of 2 (like the underlying element), but then render a cropped window on outputs of scales 1 and 2? Then this description is ambiguous.
> 
> We should rather document what happens if the scale does not match the buffer scale. Or if possible get rid of the scale and specify the `crop_rect` in `Buffer`-coordinate space?

We do not have access to the buffer space here as the `Element`-Trait does not expose the buffer scale. The intent is that downstream can just specify a rect on screen that is used for cropping, this makes it easier to crop a whole surface tree for example. I think it should be possible to move the logic to the element logic instead at construction time.

Regarding different output scales. I don't think we support that anyway, both outputs create their own element tree already as the positioning already depends on the scale.

Reply:
Ok, no, moving the logic to run-time instead of creation time is not possible, I will change the docs a bit instead.

Reply:
relaxed the scaling part and clarified the assumptions in the docs